os/hal/ports/STM32/LLD/TIMv1/hal_qei_lld.h

```diff
@@ -0,0 +1,347 @@
+#endif
+
+#if STM32_QEI_USE_TIM3 && !STM32_HAS_TIM3
+#error "TIM3 not present in the selected device"
+#endif
+
+#if STM32_QEI_USE_TIM4 && !STM32_HAS_TIM4
+#error "TIM4 not present in the selected device"
+#endif
+
+#if STM32_QEI_USE_TIM5 && !STM32_HAS_TIM5
+#error "TIM5 not present in the selected device"
+#endif
+
+#if STM32_QEI_USE_TIM8 && !STM32_HAS_TIM8
+#error "TIM8 not present in the selected device"
+#endif
+
+#if !STM32_QEI_USE_TIM1 && !STM32_QEI_USE_TIM2 &&                           \
+    !STM32_QEI_USE_TIM3 && !STM32_QEI_USE_TIM4 &&                           \
+    !STM32_QEI_USE_TIM5 && !STM32_QEI_USE_TIM8
+#error "QEI driver activated but no TIM peripheral assigned"
+#endif
+
+#if STM32_QEI_USE_TIM1 &&                                                   \
+    !OSAL_IRQ_IS_VALID_PRIORITY(STM32_QEI_TIM1_IRQ_PRIORITY)
+#error "Invalid IRQ priority assigned to TIM1"
+#endif
+
+#if STM32_QEI_USE_TIM2 &&                                                   \
+    !OSAL_IRQ_IS_VALID_PRIORITY(STM32_QEI_TIM2_IRQ_PRIORITY)
+#error "Invalid IRQ priority assigned to TIM2"
+#endif
+
+#if STM32_QEI_USE_TIM3 &&                                                   \
+    !OSAL_IRQ_IS_VALID_PRIORITY(STM32_QEI_TIM3_IRQ_PRIORITY)
+#error "Invalid IRQ priority assigned to TIM3"
+#endif
+
+#if STM32_QEI_USE_TIM4 &&                                                   \
+    !OSAL_IRQ_IS_VALID_PRIORITY(STM32_QEI_TIM4_IRQ_PRIORITY)
+#error "Invalid IRQ priority assigned to TIM4"
+#endif
+
+#if STM32_QEI_USE_TIM5 &&                                                   \
+    !OSAL_IRQ_IS_VALID_PRIORITY(STM32_QEI_TIM5_IRQ_PRIORITY)
+#error "Invalid IRQ priority assigned to TIM5"
+#endif
+
+#if STM32_QEI_USE_TIM8 &&                                                   \
+    !OSAL_IRQ_IS_VALID_PRIORITY(STM32_QEI_TIM8_IRQ_PRIORITY)
+#error "Invalid IRQ priority assigned to TIM8"
+#endif
+
+/*===========================================================================*/
+/* Driver data structures and types.                                         */
+/*===========================================================================*/
+
+/**
+ * @brief   QEI count mode.
+ */
+typedef enum {
+  QEI_MODE_QUADRATURE = 0,          /**< Quadrature encoder mode.           */
+  QEI_MODE_DIRCLOCK = 1,            /**< Direction/Clock mode.              */
+} qeimode_t;
+
+/**
+ * @brief   QEI resolution.
+ */
+typedef enum {
+  QEI_SINGLE_EDGE = 0,        /**< Count only on edges from first channel.  */
+  QEI_BOTH_EDGES = 1,         /**< Count on both edges (resolution doubles).*/
+} qeiresolution_t;
+
+/**
+ * @brief   QEI direction inversion.
+ */
+typedef enum {
+  QEI_DIRINV_FALSE = 0,             /**< Do not invert counter direction.   */
+  QEI_DIRINV_TRUE = 1,              /**< Invert counter direction.          */
+} qeidirinv_t;
+
+/**
+ * @brief   QEI counter type.
+ */
+typedef uint16_t qeicnt_t;
+
+/**
+ * @brief   QEI delta type.
+ */
+typedef int16_t qeidelta_t;

```

Comment:
Not sure but it looks like potential integer overflow. Result of uint16_t - uit16_t may be more than int16_t able to store.